camel-openmrs-fhir/src/main/java/org/openmrs/eip/fhir/routes/resources/ProcedureRouter.java

```diff
@@ -0,0 +1,79 @@
+import org.apache.camel.model.dataformat.JsonLibrary;
+import org.hl7.fhir.r4.model.CodeableConcept;
+import org.hl7.fhir.r4.model.Coding;
+import org.hl7.fhir.r4.model.Reference;
+import org.hl7.fhir.r4.model.ServiceRequest;
+import org.openmrs.eip.fhir.FhirResource;
+import org.openmrs.eip.fhir.routes.resources.dto.Order;
+import org.openmrs.eip.fhir.spring.OpenmrsRestConfiguration;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.stereotype.Component;
+
+import lombok.Setter;
+
+@Setter
+@Component
+public class ProcedureRouter extends BaseFhirResourceRouter {
+	
+	@Value("${openmrs.baseUrl}")
+	private String openmrsBaseUrl;
+	
+	@Autowired
+	private OpenmrsRestConfiguration openmrsRestConfiguration;
+	
+	ProcedureRouter() {
+		super(FhirResource.PROCEDURE);
+	}
+	
+	@Override
+	public void configure() throws Exception {
+		getCamelContext().getComponent("http", HttpComponent.class)
+		        .setHttpClientConfigurer(openmrsRestConfiguration.createHttpClientConfigurer());
+		
+		from(FhirResource.PROCEDURE.incomingUrl()).routeId("fhir-procedure-router").filter(isSupportedTable()).toD(
+		    "sql:SELECT ot.uuid as uuid from order_type ot join orders o on o.order_type_id = ot.order_type_id where o.uuid ='${exchangeProperty.event.identifier}'?dataSource=#openmrsDataSource")
+		        .filter(simple("${body[0]['uuid']} == '" + PROCEDURE_ORDER_TYPE_UUID + "'"))
+		        .log(LoggingLevel.INFO, "Processing ProcedureRouter ${exchangeProperty.event.tableName}")
+		        .toD(
+		            "sql:SELECT voided, order_action, previous_order_id FROM orders WHERE uuid = '${exchangeProperty.event.identifier}'?dataSource=#openmrsDataSource")
+		        .choice().when(simple("${exchangeProperty.event.operation} == 'd' || ${body[0]['voided']} == 1"))
+		        .setHeader(HEADER_FHIR_EVENT_TYPE, constant("d")).setBody(simple("${exchangeProperty.event.identifier}"))
+		        .to(FhirResource.PROCEDURE.outgoingUrl()).otherwise().setHeader("CamelHttpMethod", constant("GET"))
+		        .toD(openmrsBaseUrl + "/ws/rest/v1/order/${exchangeProperty.event.identifier}").unmarshal()
+		        .json(JsonLibrary.Jackson, Order.class).process(exchange -> {
+			        Order order = exchange.getIn().getBody(Order.class);
+			        exchange.getMessage().setBody(mapOrderToServiceRequest(order));
+		        }).setHeader(HEADER_FHIR_EVENT_TYPE, simple("${exchangeProperty." + PROP_EVENT_OPERATION + "}"))
+		        .to(FhirResource.PROCEDURE.outgoingUrl()).endChoice().end();
+	}
+	
+	private ServiceRequest mapOrderToServiceRequest(Order order) {
+		ServiceRequest serviceRequest = new ServiceRequest();
+		if (order.getAction().equals("DISCONTINUE")) {
+			serviceRequest.setStatus(ServiceRequest.ServiceRequestStatus.COMPLETED);
+		} else {
+			serviceRequest.setStatus(ServiceRequest.ServiceRequestStatus.ACTIVE);
+		}
```

Comment:
The mapping between OpenMRS `Order` statuses(action types) to FHIR `ServiceRequest` statuses is NOT exhaustive. OpenMRS action types include; `NEW,`  `REVISE,`  `DISCONTINUE,` and `RENEW.` First, check if there's a previous order `(order.getPreviousOrder() != null)` and then look at the action types:

-  DISCONTINUE - new order stops/replaces the previous order
-  REVISE - new order modifies/replaces the previous order
-  RENEW - new order is based on the previous order

When an order replaces another (DISCONTINUE or REVISE), In FHIR ServiceRequest set; `serviceRequest.setReplaces(...)` and when an order is renewed; `serviceRequest.setBasedOn(...)`